### PR TITLE
fix: Don't attempt to resolve artifacts if task is going to be skipped

### PR DESCRIPTION
**Pull request by simster7**
Fixes: #2167

If – once parameter resolution has been done – we determine that the task will not run, do not attempt to resolve artifacts.

### CODE REVIEW

File: workflow/controller/dag.go [hunk 1]
```diff
@@ -520,6 +520,24 @@ func (woc *wfOperationCtx) resolveDependencyReferences(dagCtx *dagContext, task
...
 
+	// If we are not executing, don't attempt to resolve any artifact references. We only check if we are executing after
+	// the initial parameter resolution, since it's likely that the "when" clause will contain parameter references.
+	proceed, err := shouldExecute(newTask.When)
+	if err != nil {
+		// If we got an error, it might be because our "when" clause contains a task-expansion parameter (e.g. {{item}}).
+		// Since we don't perform task-expansion until later and task-expansion parameters won't get resolved here,
+		// we continue execution as normal
+		if newTask.ShouldExpand() {
+			proceed = true
```
Comment (alexec) [inline workflow/controller/dag.go:531]: test?

File: workflow/controller/steps.go [hunk 2]
```diff
@@ -353,14 +353,33 @@ func (woc *wfOperationCtx) resolveReferences(stepGroup []wfv1.WorkflowStep, scop
...
 
+		// If we are not executing, don't attempt to resolve any artifact references. We only check if we are executing after
+		// the initial parameter resolution, since it's likely that the "when" clause will contain parameter references.
+		proceed, err := shouldExecute(newStep.When)
+		if err != nil {
+			// If we got an error, it might be because our "when" clause contains a task-expansion parameter (e.g. {{item}}).
+			// Since we don't perform task-expansion until later and task-expansion parameters won't get resolved here,
+			// we continue execution as normal
+			if newStep.ShouldExpand() {
+				proceed = true
```
Comment (alexec) [inline workflow/controller/steps.go:364]: test?

File: workflow/controller/dag.go (hunk 1 above)
Comment (simster7) [reply workflow/controller/dag.go:531]: Covered by an existing E2E test

File: workflow/controller/steps.go (hunk 2 above)
Comment (simster7) [reply workflow/controller/steps.go:364]: Covered by an existing E2E test

**Comment by simster7** [comment]
Back-ported to 2.7

**Pull request merged**